src/group_theory/nilpotent.lean

```diff
@@ -574,6 +576,101 @@ instance nilpotent_quotient_of_nilpotent (H : subgroup G) [H.normal] [h : is_nil
 lemma nilpotency_class_quotient_le (H : subgroup G) [H.normal] [h : is_nilpotent G] :
   group.nilpotency_class (G ⧸ H) ≤ group.nilpotency_class G := nilpotency_class_le_of_surjective _ _
 
+lemma comap_comap_center {H₁ : subgroup G} [H₁.normal] {H₂ : subgroup (G ⧸ H₁)} [H₂.normal] :
+  comap (mk' H₁) (comap (mk' H₂) (center ((G ⧸ H₁) ⧸ H₂))) =
+    comap (mk' (comap (mk' H₁) H₂)) (center (G ⧸ comap (mk' H₁) H₂)) :=
+begin
+  ext x,
+  simp only [mk'_apply, mem_comap, mem_center_iff, forall_coe],
+  apply forall_congr,
+  change ∀ (y : G), (↑↑(y * x) = ↑↑(x * y) ↔ ↑(y * x) = ↑(x * y)),
+  intro y,
+  repeat { rw [eq_iff_div_mem] },
+  simp,
+end
+
+-- This lemma helps with rewriting the subgroup, which occurs in indices
+lemma comap_center_subst {H₁ H₂ : subgroup G} [normal H₁] [normal H₂] (h : H₁ = H₂) :
+  comap (mk' H₁) (center (G ⧸ H₁)) = comap (mk' H₂) (center (G ⧸ H₂)) :=
+  by { unfreezingI { subst h, } }
+
+lemma comap_upper_central_series_quotient_center (n : ℕ) :
+  comap (mk' (center G)) (upper_central_series (G ⧸ center G) n) = upper_central_series G n.succ :=
+begin
+  induction n with n ih,
+  { simp, },
+  { let Hn := upper_central_series (G ⧸ center G) n,
+    calc comap (mk' (center G)) (upper_central_series (G ⧸ center G) n.succ)
+        = comap (mk' (center G)) (upper_central_series_step Hn)
+        : rfl
+    ... = comap (mk' (center G)) (comap (mk' Hn) (center ((G ⧸ center G) ⧸ Hn)))
+        : by rw upper_central_series_step_eq_comap_center
+    ... = comap (mk' (comap (mk' (center G)) Hn)) (center (G ⧸ (comap (mk' (center G)) Hn)))
+        : comap_comap_center
+    ... = comap (mk' (upper_central_series G n.succ)) (center (G ⧸ upper_central_series G n.succ))
+        : comap_center_subst ih
+    ... = upper_central_series_step (upper_central_series G n.succ)
+        : symm (upper_central_series_step_eq_comap_center _)
+    ... = upper_central_series G n.succ.succ
+        : rfl, },
+end
+
+lemma nilpotency_class_zero_iff_subsingleton [is_nilpotent G] :
+  group.nilpotency_class G = 0 ↔ subsingleton G :=
+by simp [group.nilpotency_class, nat.find_eq_zero, subsingleton_iff_bot_eq_top]
+
+section classical
+
+open_locale classical
+
+/-- Quotienting the `center G` reduces the nilpotency class by 1 -/
+lemma nilpotency_class_quotient_center [hH : is_nilpotent G] :
+  group.nilpotency_class (G ⧸ center G) = group.nilpotency_class G - 1 :=
```

Comment:
Does it make sense to also state the version `nilp_class G = nilp_class (G / center G) + 1`, assuming `[nontrivial G]`?

Reply:
I didn’t need it so far, but sure, I can add it.